driver-core/src/main/java/com/datastax/driver/core/policies/ErrorAwarePolicy.java

```diff
@@ -0,0 +1,529 @@
+                    }
+                } else {
+                    err.setErrorsPerSecRate(errorsPerSecHostRate);
+                }
+            }
+
+            // Update the retry periods
+            for (Map.Entry<Host, ErrorRecord> errorsRecsByHost : errorsPerSec.entrySet()) {
+                Integer nbErrorsSinceLastUpdate = errorsSinceLastUpdate.get(errorsRecsByHost.getKey());
+                if (nbErrorsSinceLastUpdate != null && nbErrorsSinceLastUpdate > 0) {
+                    // Reset the retryPeriod if there has been new errors since the last update
+                    errorsRecsByHost.getValue().setRetryPeriod(retryPeriod);
+                    continue;
+                }
+
+                // for the rest only decrement the retryPeriod
+                ErrorRecord err = errorsRecsByHost.getValue();
+                if (err.getRetryPeriod() > 0) {
+                    err.setRetryPeriod(err.getRetryPeriod() - updatePeriod);
+                }
+
+            }
+        }
+
+        @Override
+        public boolean isHostValid(Host host) {
+            ErrorRecord errorsPerSecForHost = errorsPerSec.get(host);
+            if (errorsPerSecForHost == null) {
+                // assume there hasn't been errors since startup
+                // if there's no error, it's under threshold
+                return true;
+            } else {
+                /*
+                Host is valid if :
+                - its retryPeriod is <= 0, retryPeriod > 0 means the host have had errors recently.
+                - its error rate is strictly under the threshold.
+                 */
+                return errorsPerSecForHost.getRetryPeriod() <= 0 && errorsPerSecForHost.getErrorsPerSecRate() < errorsThresholdRate;
+            }
+        }
+
+        /**
+         * Returns whether an error should be accounted by the {@link PerHostErrorTracker}.
+         * <p/>
+         * This method will be called by the {@link PerHostErrorTracker} everytime the tracker gets notified
+         * of an error. The first thing the {@link PerHostErrorTracker#notifyError} does, is call this method, if this
+         * method returns {@code true}, the error will be accounted, if it returns {@code false}, the error will
+         * be ignored.
+         * <p/>
+         * The default implementation of this method returns {@code true}, so all errors will be considered.
+         * <p/>
+         * The purpose of this method is to allow users of the {@link PerHostErrorTracker} to easily filter out
+         * errors without concerning about the implementation of the {@link PerHostErrorTracker}.
+         *
+         * @param host      the host the request was sent to.
+         * @param statement the statement that was sent.
+         * @param e         the exception the query caused.
+         * @return a {@code boolean} set to {@code true}.
+         */
+        public boolean shouldAccountForError(Host host, Statement statement, Exception e) {
```

Comment:
One thought that we've briefly discussed but thought it'd be worth putting in writing.  By default, I think we should consider not accounting for a certain set of Exceptions, errors that may not be an issue with the coordinator itself:

* ReadTimeoutException
* WriteTimeoutException
* UnavailableException
* ReadFailureException
* WriteFailureException

These errors may not necessarily mean that there is a problem with the coordinator in a lot of cases.  We could also consider doing some heuristics (like only penalizing if 0 replicas were available), but not sure if I like that idea.

Consider if we accounted for `UnavailableException`,  if a replica set cannot meet CL, we could potentially mark all hosts in the cluster down, even though some replica sets for tokens may be available.

My worry about accounting for all errors by default is that if the user uses `ErrorAwarePolicy` its likely that they don't consider this and could put their client in a bad state by using it.